### PR TITLE
feat: [CI-4569]: CLI tool to split tests

### PR DESCRIPTION
**Pull request by hemanthmantri**
An initial modified version of https://github.com/leonid-shevtsov/split_tests adapted for Harness CI

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/34553)

### CODE REVIEW

**Review by happyhebaby — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34553)* status: Some of the checks are too old. 9 file(s) to review. 0 of 1 approvals obtained. 

___
*[`product/ci/split_tests/main.go` line 188 at r1](https://reviewable.io/reviews/harness/harness-core/34553#-N5lEsiR1oq9MsDsHNzB:-N5lEsiR1oq9MsDsHNzC:bu6p1dw) ([raw file](https://github.com/harness/harness-core/blob/d44c7a430182d0a55201695c1f8ff639ce21c335/product/ci/split_tests/main.go#L188)):*
> ```Go
> 
> // Remove files that don't exist from the fileTimes map.
> func removeDeletedFiles(fileTimes map[string]float64, currentFileSet map[string]bool) {
> ```

This doesn't get used from anywhere, guess yo need to call this after fileTime map is generated?
___
*[`product/ci/split_tests/main.go` line 196 at r1](https://reviewable.io/reviews/harness/harness-core/34553#-N5lGruxDx9gkbPsj4s9:-N5lGruxDx9gkbPsj4sA:b-sw7u04) ([raw file](https://github.com/harness/harness-core/blob/d44c7a430182d0a55201695c1f8ff639ce21c335/product/ci/split_tests/main.go#L196)):*
> ```Go
> }
> 
> func addNewFiles(fileTimes map[string]float64, currentFileSet map[string]bool) {
> ```

This doesn't get used from anywhere either.

**Comment by jamie-harness** [comment]
Is there a better way than copying from another repo? Is it possible we include that one as a dependency?

**Review by happyhebaby — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34553)* status: 15 file(s) to review. 0 of 1 approvals obtained. 

___
*[`product/ci/split_tests/main_test.go` line 1 at r2](https://reviewable.io/reviews/harness/harness-core/34553#-N5pmaQF-W5Xrk8ENszk:-N5pmaQF-W5Xrk8ENszl:b-ejoga3) ([raw file](https://github.com/harness/harness-core/blob/b1dfd9b3123c7d4515d7b505a75d2075dc4ea1bc/product/ci/split_tests/main_test.go#L1)):*
> ```Go
> package main
> ```

It looks to me two more test cases could be added: 1. Test the delete logic when a fileTimeMap file isn't a current file; 2. Test missing file use average time.
It's your call, totally depends on how much testing we want to run against this functionality.
___
*[`product/ci/split_tests/main_test.go` line 15 at r2](https://reviewable.io/reviews/harness/harness-core/34553#-N5plmVI3R7dHV6iMRd-:-N5plmVI3R7dHV6iMRd0:b-829hsl) ([raw file](https://github.com/harness/harness-core/blob/b1dfd9b3123c7d4515d7b505a75d2075dc4ea1bc/product/ci/split_tests/main_test.go#L15)):*
> ```Go
> 		wantErr  bool
> 	}{
> 		{name: "Test1: 10 lines",
> ```

Just curious do we really need to test with 5000 line file? It looks to me two test files with 10 and 1000 lines would suffice.

**Review by jamie-harness — COMMENTED** [review]
Synced with Hemanth, there are a lot of customized changes to it wouldn't be good to patch as dependency

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34553#-:-N5pqS1VCqrgmZenP1eM:bp0f59z)* status: 15 file(s) to review. 0 of 1 approvals obtained.

**Review by jamie-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34553)* status: 15 file(s) to review. 0 of 1 approvals obtained. 

___
*[`BUILD.bazel` line 76 at r2](https://reviewable.io/reviews/harness/harness-core/34553#-N5pqwZ92Eh7uq5IU_9C:-N5pqwZ92Eh7uq5IU_9D:b-xujhch) ([raw file](https://github.com/harness/harness-core/blob/b1dfd9b3123c7d4515d7b505a75d2075dc4ea1bc/BUILD.bazel#L76)):*
> ```Starlark
>         "//product/ci/engine:engine_lib",
>         "//product/ci/addon:addon_lib",
> 	"//product/ci/split_tests:split_tests_lib",
> ```

I think we still use spaces instead of tabs in bazel files

File: product/ci/split_tests/main_test.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+)
+
+func Test_countLines(t *testing.T) {
+	tests := []struct {
+		name     string
+		fileName string
+		want     int
+		wantErr  bool
+	}{
+		{name: "Test1: 10 lines",
```
Comment (rutvijmehta-harness) [inline product/ci/split_tests/main_test.go:15]: Name on the next line everywhere

File: product/ci/split_tests/main.go [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+		}
+		// add file to bucket
+		buckets[minBucket] = append(buckets[minBucket], file.name)
+		bucketTimes[minBucket] += file.time
+	}
+
+	return buckets, bucketTimes
+}
+
+func processFiles(fileTimesMap map[string]float64, currentFileSet map[string]bool) {
```
Comment (rutvijmehta-harness) [inline product/ci/split_tests/main.go:197]: Description

File: product/ci/split_tests/main.go [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+
+	// Construct a map of file times {fileName: Duration}
+	fileTimesMap := make(map[string]float64)
+	if args.SplitByLineCount {
+		estimateFileTimesByLineCount(currentFileSet, fileTimesMap)
+	} else if args.UseJunitXml {
+		junit.GetFileTimesFromJUnitXML(args.JunitXmlPath, fileTimesMap)
+	}
+
+	processFiles(fileTimesMap, currentFileSet)
```
Comment (rutvijmehta-harness) [inline product/ci/split_tests/main.go:261]: Do we need to do this for linecount? 

Ideally if linecount is 0, the weight should be considered 0 and not average. For time data, we make that assumption that if the data is not present, set average.

**Review by rutvijmehta-harness — COMMENTED** [review]
LGTM overall

**Comment by rutvijmehta-harness** [comment]
*[`product/ci/split_tests/main.go` line 261 at r2](https://reviewable.io/reviews/harness/harness-core/34553#-N5996fRKZVyl247m0bx-r2-261:-N5qL_9gAH1D1sNhiFo8:b2nhb2c) ([raw file](https://github.com/harness/harness-core/blob/b1dfd9b3123c7d4515d7b505a75d2075dc4ea1bc/product/ci/split_tests/main.go#L261)):*
<details><summary><i>Previously, hemanthmantri (Hemanth Mantri) wrote…</i></summary><blockquote>

ProcessFiles is to basically make sure the map has the same files as currently known files. Independent of what time units are (length, size, duration). If the line count is 0 we will get it via countLines() method and we don't need to add it. We can special case and handle them differently but prefer to keep it generic.
</blockquote></details>

Ack

**Review by rutvijmehta-harness — APPROVED** [review]
@hemanthmantri there were issues in develop which is why you have Codeformat failing (some errors are from your bazel changes too). You'll have to rebase to get past the unrelated errors

**Review by hemanthmantri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34553)* status: 15 file(s) to review. 8. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`product/ci/split_tests/main.go` line 86 at r2](https://reviewable.io/reviews/harness/harness-core/34553#-N5996fRKZVyl247m0bx-r2-86:-N5qYbkI876YAKfMzbxf:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/b1dfd9b3123c7d4515d7b505a75d2075dc4ea1bc/product/ci/split_tests/main.go#L86)):*
<details><summary><i>Previously, rutvijmehta-harness (Rutvij Mehta) wrote…</i></summary><blockquote>

Ack. You can put a comment there so it's not confusing for someone who's reading for the first time.
</blockquote></details>

Done.

**Review by rutvijmehta-harness — APPROVED** [review]
> Reviewed 5 of 9 files at r1, 10 of 10 files at r4, 8 of 8 files at r5.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34553)* status: 6 file(s) to review. 6. unresolved discussions.

**Review by rutvijmehta-harness — APPROVED** [review]
> Reviewed 5 of 5 files at r6, 1 of 1 files at r7, 2 of 2 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34553)* status: 5. unresolved discussions.

**Review by rutvijmehta-harness — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34553#-:-N5rih0T55mMr6Rq1-6d:bnfp4nl)* status: 5. unresolved discussions.

**Review by happyhebaby — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34553)* status: 1. unresolved discussions. 

___
*[`product/ci/split_tests/main.go` line 188 at r1](https://reviewable.io/reviews/harness/harness-core/34553#-N5lEsiR1oq9MsDsHNzB:-N5sbH534EhP2IBViDvV:b-l255k6) ([raw file](https://github.com/harness/harness-core/blob/d44c7a430182d0a55201695c1f8ff639ce21c335/product/ci/split_tests/main.go#L188)):*
<details><summary><i>Previously, hemanthmantri (Hemanth Mantri) wrote…</i></summary><blockquote>

This is used in main method after processFiles() is done.
</blockquote></details>

I referred to removeDeletedFiles, which was already removed in a later change.



_Code quote:_
```Go
removeDeletedFiles
```
___
*[`product/ci/split_tests/main.go` line 196 at r1](https://reviewable.io/reviews/harness/harness-core/34553#-N5lGruxDx9gkbPsj4s9:-N5sbjl81tuin5cg1Z3f:b-6l7uac) ([raw file](https://github.com/harness/harness-core/blob/d44c7a430182d0a55201695c1f8ff639ce21c335/product/ci/split_tests/main.go#L196)):*
<details><summary><i>Previously, hemanthmantri (Hemanth Mantri) wrote…</i></summary><blockquote>

Used in the main() method.
</blockquote></details>

I referred to addNewFiles, which was already removed in a later change.

**Review by hemanthmantri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34553)* status: 3. unresolved discussions. 

___
*[`product/ci/split_tests/main.go` line 188 at r1](https://reviewable.io/reviews/harness/harness-core/34553#-N5lEsiR1oq9MsDsHNzB:-N5sswO34yvA3Ae1MuiT:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/d44c7a430182d0a55201695c1f8ff639ce21c335/product/ci/split_tests/main.go#L188)):*
<details><summary><i>Previously, happyhebaby (Xun He) wrote…</i></summary><blockquote>

I referred to removeDeletedFiles, which was already removed in a later change.
</blockquote></details>

Done.
___
*[`product/ci/split_tests/main.go` line 196 at r1](https://reviewable.io/reviews/harness/harness-core/34553#-N5lGruxDx9gkbPsj4s9:-N5ssx7S811eb2GYhfbJ:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/d44c7a430182d0a55201695c1f8ff639ce21c335/product/ci/split_tests/main.go#L196)):*
<details><summary><i>Previously, happyhebaby (Xun He) wrote…</i></summary><blockquote>

I referred to addNewFiles, which was already removed in a later change.
</blockquote></details>

Done.

**Review by happyhebaby — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34553)* status: 1. unresolved discussions.

**Review by hemanthmantri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34553)* status: 1. unresolved discussions. 

___
*[`product/ci/split_tests/main_test.go` line 15 at r2](https://reviewable.io/reviews/harness/harness-core/34553#-N5n_9IiD-_NWGll41X5-r2-15:-N5ucDf9683E0dReSN_x:by2y0wt) ([raw file](https://github.com/harness/harness-core/blob/b1dfd9b3123c7d4515d7b505a75d2075dc4ea1bc/product/ci/split_tests/main_test.go#L15)):*
<details><summary><i>Previously, hemanthmantri (Hemanth Mantri) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Can you resolve this please?

**Review by rutvijmehta-harness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/34553)* status: :shipit: complete! ✓

**Pull request merged**